authenticatorbridge/src/main/java/com/bitwarden/authenticatorbridge/manager/AuthenticatorBridgeManagerImpl.kt

```diff
@@ -0,0 +1,189 @@
+import com.bitwarden.authenticatorbridge.provider.AuthenticatorBridgeCallbackProvider
+import com.bitwarden.authenticatorbridge.provider.StubAuthenticatorBridgeCallbackProvider
+import com.bitwarden.authenticatorbridge.provider.SymmetricKeyStorageProvider
+import com.bitwarden.authenticatorbridge.util.decrypt
+import com.bitwarden.authenticatorbridge.util.toFingerprint
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+
+private const val AUTHENTICATOR_BRIDGE_SERVICE_CLASS =
+    "com.x8bit.bitwarden.data.platform.service.AuthenticatorBridgeService"
+
+/**
+ * Default implementation of [AuthenticatorBridgeManager].
+ *
+ * @param applicationContext The Context that will be used to bind to AuthenticatorBridgeService. Should be an
+ * application context.
+ * @param connectionType Specifies which build variant to connect to.
+ * @param symmetricKeyStorageProvider Provides access to local storage of the symmetric encryption
+ * key.
+ * @param callbackProvider Provides a way to construct a service callback that can be mocked in
+ * tests.
+ * @param processLifecycleOwner Lifecycle owner that is used to listen for start/stop
+ * lifecycle events.
+ */
+class AuthenticatorBridgeManagerImpl(
+    private val applicationContext: Context,
+    private val connectionType: AuthenticatorBridgeConnectionType,
+    private val symmetricKeyStorageProvider: SymmetricKeyStorageProvider,
+    callbackProvider: AuthenticatorBridgeCallbackProvider = StubAuthenticatorBridgeCallbackProvider(),
+    processLifecycleOwner: LifecycleOwner = ProcessLifecycleOwner.get(),
+) : AuthenticatorBridgeManager, DefaultLifecycleObserver {
+
+    /**
+     * Keep track of whether the service is bound. Useful for making sure we don't call
+     * [Context.unbindService] on a service that isn't bound.
+     */
+    private var isBound = false
+
+    /**
+     * Main AuthenticatorBridgeService access point.
+     */
+    private var bridgeService: IAuthenticatorBridgeService? = null
+
+    /**
+     * Internal state of [accountSyncStateFlow].
+     */
+    private
+    val mutableSharedAccountsStateFlow =
+        MutableStateFlow<AccountSyncState>(AccountSyncState.Loading)
+
+    /**
+     * Callback registered with AuthenticatorBridgeService.
+     */
+    private val authenticatorBridgeCallback = callbackProvider.getCallback(::onAccountsSync)
+
+    /**
+     * Service connection that listens for connected and disconnected service events.
+     */
+    private val bridgeServiceConnection = object : ServiceConnection {
+        override fun onServiceConnected(name: ComponentName, service: IBinder) {
+            bridgeService = IAuthenticatorBridgeService.Stub.asInterface(service)
+
+            // TODO: Add check for version mismatch between client and server SDKs:
+            // TODO: https://livefront.atlassian.net/browse/BITAU-72
```

Comment:
Honestly based on our designs for errors, it's possible we don't even need to worry about this right now. Just something to keep in mind.

Reply:
Yeah I've been noodling on this a bit.